ir/src/passes/translate.rs

```diff
@@ -1,46 +1,80 @@
-use air_parser::{ast, LexicalScope};
+use air_parser::{ast, symbols, LexicalScope, SemanticAnalysisError};
 use air_pass::Pass;
 
-use miden_diagnostics::{DiagnosticsHandler, Severity, Span, Spanned};
+use miden_diagnostics::{DiagnosticsHandler, SourceSpan, Spanned};
 
 use crate::{graph::NodeIndex, ir::*, CompileError};
 
-pub struct AstToAir<'a> {
+pub struct AstToMir<'a> {
     diagnostics: &'a DiagnosticsHandler,
 }
-impl<'a> AstToAir<'a> {
+impl<'a> AstToMir<'a> {
     /// Create a new instance of this pass
     #[inline]
     pub fn new(diagnostics: &'a DiagnosticsHandler) -> Self {
         Self { diagnostics }
     }
 }
-impl<'p> Pass for AstToAir<'p> {
+impl<'p> Pass for AstToMir<'p> {
     type Input<'a> = ast::Program;
-    type Output<'a> = Air;
+    type Output<'a> = Mir;
     type Error = CompileError;
 
     fn run<'a>(&mut self, program: Self::Input<'a>) -> Result<Self::Output<'a>, Self::Error> {
-        let mut air = Air::new(program.name);
+        let mut mir = Mir::new(program.name);
+
+        //TODO MIR: Implement AST > MIR lowering
+        // 1. Start from the previous lowering from AST to AIR
+        // 2. Understand what changes when starting from an unoptimized AST
+        // (with no constant prop and no inlining)
+        // 3. Implement the needed changes
 
         let random_values = program.random_values;
         let trace_columns = program.trace_columns;
         let boundary_constraints = program.boundary_constraints;
         let integrity_constraints = program.integrity_constraints;
 
-        air.trace_segment_widths = trace_columns.iter().map(|ts| ts.size as u16).collect();
-        air.num_random_values = random_values.as_ref().map(|rv| rv.size as u16).unwrap_or(0);
-        air.periodic_columns = program.periodic_columns;
-        air.public_inputs = program.public_inputs;
+        mir.trace_segment_widths = trace_columns.iter().map(|ts| ts.size as u16).collect();
+        mir.num_random_values = random_values.as_ref().map(|rv| rv.size as u16).unwrap_or(0);
+        mir.periodic_columns = program.periodic_columns;
+        mir.public_inputs = program.public_inputs;
 
-        let mut builder = AirBuilder {
+        let mut builder = MirBuilder {
             diagnostics: self.diagnostics,
-            air: &mut air,
+            mir: &mut mir,
             random_values,
             trace_columns,
             bindings: Default::default(),
         };
 
+        // Insert placeholders nodes for future Operation::Definition (needed for function bodies to call other functions)
```

Comment:
Could you elaborate a bit on what the placeholder op is and why it is needed? Again, i suspect the representation I outlined above makes this unnecessary, but would like to make sure I understand the problem you ran into.

Reply:
The PlaceHolder Operation variant is currently used to obtain the NodeIndex before the node is fully defined to help with incremental graph edits. Main caveat is that we might have to rework the uniquing strategy.
Should be possible to do without this node type with a slight rework of how we insert nodes.

Reply:
Ah, makes sense. The approach I've taken with this in the past, and which I've seen in other similar data structures, e.g. `cranelift_entity::PrimaryMap<K, V>`, have an underlying vector of `Option<V>`, or require `V` be default-constructible, so that:

* You can allocate indices before storing anything at that slot while ensuring that all indices are valid for the underlying vector, i.e. given an index, they can simply do `items.resize_with(index + 1, || None)` or `items.resize_with(index + 1, Default::default)`, to ensure that the index is valid, and returns something sensible. I think the `Option<V>` approach works best, since you can actually determine if a slot is occupied or not, but both options let you allocate indices before you construct the actual `V`.
* You can reuse slots, assuming you use the `Option<V>` representation, by maintaining the next available index (or a free list of them, depending on how you "release" nodes). You then check if you have a slot you can reuse when a new index is requested.
* The primary API can be basically `fn push(&mut self, node: V) -> K`, i.e. you store the node and get back its handle, but if you need to store the key in the node, you need to use two operations: `fn alloc_key(&mut self) -> K` and `unsafe fn insert(&mut self, index: K, node: V);`. The latter is marked unsafe since the index must be valid, and allocated by `alloc_key`, though it isn't strictly speaking an unsafe operation in terms of memory safety.

In any case, using that approach is definitely preferable to needing a special placeholder variant in your node type that you then need to handle everywhere even though you know it will never appear in practice.